opt: Add Presentation physical property

andy-kimball opened the pull request:
This PR adds support for Presentation, the first physical property.
Presentation specifies the naming, membership (including duplicates), and
order of result columns that are required of or provided by an operator.
The Presentation property can be provided any relational operator, and
will only be required of the root of an expression tree. For this reason,
Presentation is a fairly simple physical property.

Future physical props will require more support from the optimizer in
order to work. In particular, the optimizer will soon get support for
maintaining the lowest cost expression for a particular set of physical
props. Until then, the optimizer produces a hardcoded ExprView that
traverses the normalized expression tree rather than the lowest cost
tree.

In order to test the new property, this PR also contains a new optimizer
rule which eliminates a Project operator if it projects the same set of
columns as its input. This eliminates many top-level Project operators
and ensures that the Presentation property still produces the correctly
labeled list of output columns.

Release note: None

andy-kimball committed:
opt: Add Presentation physical property

This PR adds support for Presentation, the first physical property.
Presentation specifies the naming, membership (including duplicates), and
order of result columns that are required of or provided by an operator.
The Presentation property can be provided any relational operator, and
will only be required of the root of an expression tree. For this reason,
Presentation is a fairly simple physical property.

Future physical props will require more support from the optimizer in
order to work. In particular, the optimizer will soon get support for
maintaining the lowest cost expression for a particular set of physical
props. Until then, the optimizer produces a hardcoded ExprView that
traverses the normalized expression tree rather than the lowest cost
tree.

In order to test the new property, this PR also contains a new optimizer
rule which eliminates a Project operator if it projects the same set of
columns as its input. This eliminates many top-level Project operators
and ensures that the Presentation property still produces the correctly
labeled list of output columns.

Release note: None

andy-kimball commented on pkg/sql/opt/exec/execbuilder/testdata/project:
```diff
@@ -113,21 +113,16 @@ SELECT u + v FROM (SELECT x + 3, y + 10 FROM t.a) AS foo(u, v)
 exec-explain
 SELECT x, x, y, x FROM t.a
```
@RaduBerinde, this PR breaks this test. I added a TODO in `builder.go` for you to add support for the new Presentation property, which will fix this test.

RaduBerinde commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Review status: 0 of 29 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/exec/execbuilder/testdata/project, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23037#-L65TT_yvqDZnJwc1yJ3-r1-114:-L66v3wwkXhD3Y1dpTBb:b53j4x2) ([raw file](https://github.com/cockroachdb/cockroach/blob/431306b9a187d18fe90af55bbdcc5c7976c5e7d7/pkg/sql/opt/exec/execbuilder/testdata/project#L114)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

@RaduBerinde, this PR breaks this test. I added a TODO in `builder.go` for you to add support for the new Presentation property, which will fix this test.
</blockquote></details>

Sure, sounds good. Should be an easy fix. But please comment out the tests for now instead of having incorrect results in there.

---

*[pkg/sql/opt/xform/rules/project.opt, line 8 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23037#-L68BZfZT8NRanEZI3Rk:-L68BZfZT8NRanEZI3Rl:b-9v6tim) ([raw file](https://github.com/cockroachdb/cockroach/blob/abec16b1d0b09d2d93d222a9f2d0188624e0fb66/pkg/sql/opt/xform/rules/project.opt#L8)):*
> ```
> # EliminateProject discards a Project operator which is not adding or removing
> # columns, but instead is just reordering existing columns.
> [EliminateProject, Normalize]
> ```

The way these rules look is great, very self-documenting!

---

*[pkg/sql/opt/xform/testdata/rules/project, line 41 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23037#-L68C1fOrsXYVUiQCO4H:-L68C1fOrsXYVUiQCO4I:b4bvpsq) ([raw file](https://github.com/cockroachdb/cockroach/blob/abec16b1d0b09d2d93d222a9f2d0188624e0fb66/pkg/sql/opt/xform/testdata/rules/project#L41)):*
> ```
>  └── columns: alias1:float:null:2 x:int:1 alias1:float:null:2 x:int:1
> 
> # Added column (should not eliminate).
> ```

[nit] projection should not be eliminated (same below)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23037#-:-L68C8aMtVkUAe9KD7CX:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 29 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/exec/execbuilder/testdata/project, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23037#-L65TT_yvqDZnJwc1yJ3-r1-114:-L68OGSBTROvm0Hx0hOy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/431306b9a187d18fe90af55bbdcc5c7976c5e7d7/pkg/sql/opt/exec/execbuilder/testdata/project#L114)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Sure, sounds good. Should be an easy fix. But please comment out the tests for now instead of having incorrect results in there.
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/testdata/rules/project, line 41 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23037#-L68C1fOrsXYVUiQCO4H:-L68Ob_mxwZpc959k92I:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/abec16b1d0b09d2d93d222a9f2d0188624e0fb66/pkg/sql/opt/xform/testdata/rules/project#L41)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] projection should not be eliminated (same below)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23037)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice!

---

Reviewed 25 of 29 files at r1, 2 of 2 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/exec/execbuilder/testdata/project, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23037#-L65TT_yvqDZnJwc1yJ3-r1-114:-L6HwFkAB980ELnKKHDZ:b-cady3) ([raw file](https://github.com/cockroachdb/cockroach/blob/431306b9a187d18fe90af55bbdcc5c7976c5e7d7/pkg/sql/opt/exec/execbuilder/testdata/project#L114)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Looks like the commented results are wrong - you should probably discard this change in git and then comment out the test case

---

*[pkg/sql/opt/opt/physical_props.go, line 51 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23037#-L6HxDL3R7OX0okDoZ8r:-L6HxDL3R7OX0okDoZ8s:br21p50) ([raw file](https://github.com/cockroachdb/cockroach/blob/44e6a57d328f82c66459781fcf8cb8de1248b2e1/pkg/sql/opt/opt/physical_props.go#L51)):*
> ```Go
> // PhysicalProps are interesting characteristics of an expression that impact
> // its layout, presentation, or location, but not its logical content. Examples
> // include row order, column naming, and data distribution. Physical properties
> ```

What is data distribution?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23037#-:-L6I3g0L_Dr3zDj7E8Vd:b-8gk8uj)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 9 of 32 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/opt/exec/execbuilder/testdata/project, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23037#-L65TT_yvqDZnJwc1yJ3-r1-114:-L6IMqLSm6z7y55SHSJ-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/431306b9a187d18fe90af55bbdcc5c7976c5e7d7/pkg/sql/opt/exec/execbuilder/testdata/project#L114)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Looks like the commented results are wrong - you should probably discard this change in git and then comment out the test case
</blockquote></details>

Done.

---

*[pkg/sql/opt/opt/physical_props.go, line 51 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23037#-L6HxDL3R7OX0okDoZ8r:-L6IMqyjQjwpYIwbo9Kc:b-gw1uks) ([raw file](https://github.com/cockroachdb/cockroach/blob/44e6a57d328f82c66459781fcf8cb8de1248b2e1/pkg/sql/opt/opt/physical_props.go#L51)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

What is data distribution?
</blockquote></details>

Added same parenthetical that I added to `doc.go`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23037)*
<!-- Sent from Reviewable.io -->
